File: src/core/provider.ts
/**
 * Groklets Provider Abstraction — Model-agnostic LLM provider interface.
 *
 * The provider layer ensures the core engine has zero dependency on
 * any specific LLM vendor. Providers are fully interchangeable and
 * agents bind to them declaratively via config.
 */

import { createLogger } from "../utils/logger.js";

const log = createLogger("Provider");

export enum Role {
    SYSTEM = "system",
    USER = "user",
    ASSISTANT = "assistant",
    TOOL = "tool",
}

export interface Message {
    role: Role;
    content: string;
    name?: string;
    toolCalls?: ToolCall[];
    toolCallId?: string;
    metadata?: Record<string, unknown>;
}

export interface ToolCall {
    id: string;
    type: string;
    function: { name: string; arguments: string };
}

export interface CompletionResponse {
    message: Message;
    finishReason: string;
    usage: { promptTokens: number; completionTokens: number; totalTokens: number };
    rawResponse?: unknown;
    model: string;
}

export interface ProviderConfig {
    apiKey?: string;
    model?: string;
    baseUrl?: string;
    temperature?: number;
    maxTokens?: number;
    timeout?: number;
    maxRetries?: number;
    extra?: Record<string, unknown>;
}

export interface ToolDefinition {
    type: "function";
    function: {
        name: string;
        description?: string;
        parameters?: Record<string, unknown>;
    };
}

/**
 * Every provider must implement this interface.
 * The engine interacts exclusively through it.
 */
export interface ProviderBase {
    readonly name: string;
    readonly config: ProviderConfig;
    complete(messages: Message[], tools?: ToolDefinition[], overrides?: Partial<ProviderConfig>): Promise<CompletionResponse>;
    stream?(messages: Message[], tools?: ToolDefinition[], overrides?: Partial<ProviderConfig>): AsyncIterable<string>;
    healthCheck?(): Promise<boolean>;
}

/**
 * Provider factory function — alternative to class-based registration.
 * Users can register a simple function that returns a ProviderBase.
 */
export type ProviderFactory = (config: ProviderConfig) => ProviderBase;

/**
 * Provider Registry — agents reference providers by name.
 *
 * Supports:
 *   - Built-in types (openai, anthropic, google, xai)
 *   - Custom provider classes (registerClass)
 *   - Custom provider factories (registerFactory)
 *   - Provider instances (registerInstance)
 *   - OpenAI-compatible fallback (any unknown type with a base_url)
 */
export class ProviderRegistry {
    private providers = new Map<string, ProviderBase>();
    private factories = new Map<string, ProviderFactory>();
    private classes = new Map<string, new (config: ProviderConfig) => ProviderBase>();

    /** Register a provider class by type name. */
    registerClass(name: string, cls: new (config: ProviderConfig) => ProviderBase): void {
        this.classes.set(name, cls);
        log.debug(`Registered provider class: ${name}`);
    }

    /** Register a provider factory function by type name. */
    registerFactory(name: string, factory: ProviderFactory): void {
        this.factories.set(name, factory);
        log.debug(`Registered provider factory: ${name}`);
    }

    /** Register a ready-to-use provider instance. */
    registerInstance(name: string, provider: ProviderBase): void {
        this.providers.set(name, provider);
        log.debug(`Registered provider: ${name} → ${provider.name}`);
    }

    /** Get a provider by name. */
    get(name: string): ProviderBase {
        const p = this.providers.get(name);
        if (p) return p;
        throw new Error(`Provider not found: '${name}'. Available: ${this.available.join(", ")}`);
    }

    /**
     * Create a provider from a registered type.
     * If the type is unknown but config has a baseUrl, falls back to
     * OpenAI-compatible mode (works with Ollama, LM Studio, vLLM,
     * Together, Groq, Mistral, etc.).
     */
    create(typeName: string, instanceName: string, config: ProviderConfig): ProviderBase {
        // 1. Check factories first
        const factory = this.factories.get(typeName);
        if (factory) {
            const instance = factory(config);
            this.providers.set(instanceName, instance);
            return instance;
        }

        // 2. Check registered classes
        const Cls = this.classes.get(typeName);
        if (Cls) {
            const instance = new Cls(config);
            this.providers.set(instanceName, instance);
            return instance;
        }

        // 3. Unknown type — if it has a baseUrl, assume OpenAI-compatible
        if (config.baseUrl) {
            const openaiCls = this.classes.get("openai");
            if (openaiCls) {
                log.info(`Unknown provider type '${typeName}' — using OpenAI-compatible mode with baseUrl: ${config.baseUrl}`);
                const instance = new openaiCls(config);
                this.providers.set(instanceName, instance);
                return instance;
            }
        }

        throw new Error(
            `Unknown provider type: '${typeName}'. ` +
            `Register it with registerClass() or registerFactory(), ` +
            `or set base_url for OpenAI-compatible mode. ` +
            `Built-in types: ${[...this.classes.keys()].join(", ")}`
        );
    }

    /** List all available provider names. */
    get available(): string[] {
        return [...new Set([...this.providers.keys()])];
    }

    /** List all registered provider type names. */
    get registeredTypes(): string[] {
        return [...new Set([...this.classes.keys(), ...this.factories.keys()])];
    }

    has(name: string): boolean {
        return this.providers.has(name);
    }

    hasType(typeName: string): boolean {
        return this.classes.has(typeName) || this.factories.has(typeName);
    }
}

File: swarmx/core/provider.py
"""
SwarmX Provider Abstraction — Model-agnostic LLM provider interface.

Adapted from OpenClaw's multi-model support pattern. The provider layer
ensures the core engine has zero dependency on any specific LLM vendor.
Providers are interchangeable and agents bind to them declaratively.
"""

from __future__ import annotations

import logging
from abc import ABC, abstractmethod
from dataclasses import dataclass, field
from enum import Enum
from typing import Any, AsyncIterator

logger = logging.getLogger("swarmx.provider")


class Role(Enum):
    """Standard message roles across all providers."""

    SYSTEM = "system"
    USER = "user"
    ASSISTANT = "assistant"
    TOOL = "tool"


@dataclass
class Message:
    """
    A provider-agnostic message.

    All providers consume and produce Message objects, enabling
    seamless provider switching without touching agent logic.
    """

    role: Role
    content: str
    name: str | None = None
    tool_calls: list[dict[str, Any]] | None = None
    tool_call_id: str | None = None
    metadata: dict[str, Any] = field(default_factory=dict)


@dataclass
class CompletionResponse:
    """
    Standardized response from any provider.

    Wraps the provider-specific response into a uniform format
    that the engine and agents can consume without knowing which
    provider generated it.
    """

    message: Message
    finish_reason: str = "stop"
    usage: dict[str, int] = field(default_factory=dict)
    raw_response: Any = None  # Original provider response for advanced access
    model: str = ""


@dataclass
class ProviderConfig:
    """Configuration for a provider instance."""

    api_key: str = ""
    model: str = ""
    base_url: str | None = None
    temperature: float = 0.7
    max_tokens: int = 4096
    timeout: float = 60.0
    extra: dict[str, Any] = field(default_factory=dict)


class ProviderBase(ABC):
    """
    Abstract base class for all LLM providers.

    Every provider must implement `complete()` and optionally `stream()`.
    The engine interacts exclusively through this interface.
    """

    def __init__(self, config: ProviderConfig) -> None:
        self.config = config
        self._name: str = self.__class__.__name__

    @property
    def name(self) -> str:
        """Human-readable provider name."""
        return self._name

    @abstractmethod
    async def complete(
        self,
        messages: list[Message],
        tools: list[dict[str, Any]] | None = None,
        **kwargs: Any,
    ) -> CompletionResponse:
        """
        Send a list of messages and return a completion.

        Args:
            messages: Conversation history as Message objects.
            tools: Optional tool/function definitions the model can call.
            **kwargs: Provider-specific overrides.

        Returns:
            A CompletionResponse with the model's output.
        """
        ...

    async def stream(
        self,
        messages: list[Message],
        tools: list[dict[str, Any]] | None = None,
        **kwargs: Any,
    ) -> AsyncIterator[str]:
        """
        Stream a completion token-by-token.

        Default implementation falls back to non-streaming `complete()`.
        Providers can override for true streaming support.
        """
        response = await self.complete(messages, tools, **kwargs)
        yield response.message.content

    async def health_check(self) -> bool:
        """Verify the provider is reachable and the API key is valid."""
        try:
            test_msg = [Message(role=Role.USER, content="ping")]
            await self.complete(test_msg)
            return True
        except Exception:
            logger.warning("Health check failed for %s", self.name)
            return False

    def __repr__(self) -> str:
        return f"<{self.name} model={self.config.model}>"


class ProviderRegistry:
    """
    Registry of available providers.

    Providers register themselves by name, and agents reference
    providers declaratively by name in their configuration.
    This pattern is adapted from OpenClaw's channel adapter registry.
    """

    def __init__(self) -> None:
        self._providers: dict[str, ProviderBase] = {}
        self._factory: dict[str, type[ProviderBase]] = {}

    def register_class(self, name: str, provider_class: type[ProviderBase]) -> None:
        """Register a provider class for lazy instantiation."""
        self._factory[name] = provider_class
        logger.debug("Provider class registered: %s", name)

    def register_instance(self, name: str, provider: ProviderBase) -> None:
        """Register a pre-configured provider instance."""
        self._providers[name] = provider
        logger.debug("Provider instance registered: %s → %s", name, provider)

    def get(self, name: str) -> ProviderBase:
        """Retrieve a provider by name."""
        if name in self._providers:
            return self._providers[name]
        raise KeyError(f"Provider not found: '{name}'. Available: {list(self.available)}")

    def create(self, name: str, config: ProviderConfig) -> ProviderBase:
        """Create a provider instance from a registered class."""
        if name not in self._factory:
            raise KeyError(
                f"Provider class not registered: '{name}'. "
                f"Available: {list(self._factory.keys())}"
            )
        provider = self._factory[name](config)
        self._providers[name] = provider
        return provider

    @property
    def available(self) -> list[str]:
        """List available provider names."""
        return list(set(list(self._providers.keys()) + list(self._factory.keys())))

    def __contains__(self, name: str) -> bool:
        return name in self._providers or name in self._factory
